### PR TITLE
207 postgres nested conditions

### LINKED ISSUE #207: Postgres query nested conditions
Postgres store needs to support the "reverse lookup" case - i.e. querying by nested conditions.

Use case:
```ruby
class Person < WCC::Contentful::Model::Person
  def teams
    ::Team.find_all members: { id: id }
  end
```

### PR DESCRIPTION
**Pull request by gburgett**
Support Join queries in the postgres store
Reorganize the Query interface to normalize query parsing

fixes #207

### CODE REVIEW

**Comment by gburgett** [comment]
Still TODO: fix the Memory store or indicate it's reduced capabilities in documentation...

File: wcc-contentful/lib/wcc/contentful/store/base.rb [hunk 1]
```diff
@@ -196,8 +201,35 @@ def resolve_link(val)
...
+        if value.keys.length == 1
+          k, v = value.first
+          return _apply([field, k].join('.'), v, context) if k == 'id' || (k == 'sys' && v == 'id')
+        end
+
+        self_join(field, value, context)
+      end
+
+      def self_join(join_on, conditions, context)
+        apply(conditions.transform_keys { |k| [join_on, k].join('.') }, context)
```
Comment (gburgett) [inline wcc-contentful/lib/wcc/contentful/store/base.rb]: This is unused - I really ought to just raise an error here

File: wcc-contentful/lib/wcc/contentful/store/postgres_store.rb [hunk 2]
```diff
@@ -54,131 +56,167 @@ def find(key, **_options)
...
+
+    def _join(join_path, expectation_path, expected, params, joins)
+      join_table_alias = push_join(join_path, joins)
+      " AND #{join_table_alias}.data->#{quote_parameter_path(expectation_path)}" \
+        " ? $#{push_param(expected, params)}::text"
+    end
 
-        params = @params.dup
+    def push_join(path, joins)
+      table_alias = "s#{joins.length}"
```
Comment (jpowell) [inline wcc-contentful/lib/wcc/contentful/store/postgres_store.rb]: 👍

File: wcc-contentful/lib/wcc/contentful/store/query.rb [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+        context_locale = context[:locale] if context.present?
+        context_locale ||= 'en-US'
+
+        rev_path = path.reverse
+        new_path = []
+
+        current_tuple = []
+        current_locale_was_inferred = false
+        until rev_path.empty? && current_tuple.empty?
+          raise ArgumentError, "Query too complex: #{path.join('.')}" if new_path.length > 7
```
Comment (jpowell) [inline wcc-contentful/lib/wcc/contentful/store/query.rb]: is this enforcing no more than one layer of nested querying?

Comment (gburgett) [reply wcc-contentful/lib/wcc/contentful/store/query.rb]: Yep

**Pull request merged**